book/src/template_expansion.md

```diff
@@ -0,0 +1,387 @@
+let y = x + 1;
+::askama::Result::Ok(())
+```
+
+### Variables usage
+
+By default, variables will be fields from the type on which the `askama::Template`
+trait is implemented:
+
+```jinja
+{{ y }}
+```
+
+will generate:
+
+```rust
+writer
+    .write_fmt(
+        format_args!(
+            "{0}",
+            &::askama::MarkupDisplay::new_unsafe(&(self.y), ::askama::Html),
+        ),
+    )?;
+::askama::Result::Ok(())
+```
+
+This is why if the variable is undefined, it won't work with askama and why
+we can't check if a variable is defined or not.
+
+However, you can still access constants and statics by using paths. Let's say
+you have in your Rust code:
+
+```rust
+const FOO: u32 = 0;
+```
+
+Then you can use them in your template by referring to them with a path:
+
+```jinja
+{{ crate::FOO }}{{ super::FOO }}
+```
+
+It will generate:
+
+```rust
+writer
+    .write_fmt(
+        format_args!(
+            "{0}{1}",
+            &::askama::MarkupDisplay::new_unsafe(&(crate::FOO), ::askama::Html),
+            &::askama::MarkupDisplay::new_unsafe(&(super::FOO), ::askama::Html),
+        ),
+    )?;
+::askama::Result::Ok(())
+```
+
+You can also access items from your current type as well using `Self::`, it'll
+use the same logic.
+
+## Control blocks
+
+### if/else
+
+There is an interesting twist for `if/else` conditions. Let's check this code:
+
+```jinja
+{% if x == "a" %}
+gateau
+{% else %}
+tarte
+{% endif %}
+```
+
+It will generate:
+
+```rust
+if *(&(self.x == "a") as &bool) {
+    writer.write_str("gateau")?;
+} else {
+    writer.write_str("tarte")?;
+}
+::askama::Result::Ok(())
+```
+
+Very much as expected except for the `&(self.x == "a") as &bool`. It allows to
+fallback to a boolean as much as possible, but it also explains why you can't
+do:
+
+```jinja
+{% set x = "a" %}
+{% if x %}
+    {{ x }}
+{% endif %}
+```
+
+Because it fail to compile because:
+
+```console
+error[E0605]: non-primitive cast: `&&str` as `&bool`
+```
+
+### if let
+
+```jinja
+{% if let Some(x) = x %}
+    {{ x }}
+{% endif %}
+```
+
+will generate:
+
+```rust
+if let Some(x) = &(self.x) {
+    writer
+        .write_fmt(
+            format_args!(
+                "{0}",
+                &::askama::MarkupDisplay::new_unsafe(&(x), ::askama::Html),
+            ),
+        )?;
+}
+```
+
+### Loops
+
+```html
+{% for user in users %}
+    {{ user }}
+{% endfor %}
+```
+
+will generate:
+
+```rust
+{
+    let mut _did_loop = false;
+    let _iter = (&self.users).into_iter();
+    for (user, _loop_item) in ::askama::helpers::TemplateLoop::new(_iter) {
+        _did_loop = true;
+        writer
+            .write_fmt(
+                format_args!(
+                    "\n    {0}\n",
+                    &::askama::MarkupDisplay::new_unsafe(&(user), ::askama::Html),
+                ),
+            )?;
+    }
+    if !_did_loop {}
+}
+::askama::Result::Ok(())
+```
+
+Interestingly enough, there is a `_did_loop` variable. It is used in case your
```

Comment:
This seems to conflate two different concepts:

* `if`-based loop filters
* The `else` branch, which should presumably also work for unfiltered iterators that are empty?

(We should arguably not generate the `if !did_loop {}` -- and maybe even if the `did_loop` variable -- if the `for` block didn't have an `else` clause?)

Reply:
I was planning to send a follow-up to clean things up. :)

Reply:
> The `else` branch, which should presumably also work for unfiltered iterators that are empty?

Do you have an example by any chance?

Reply:
> > The `else` branch, which should presumably also work for unfiltered iterators that are empty?
> 
> Do you have an example by any chance?

Ping @djc :)

Reply:
You wrote "It is used in case your `for` loop has a `if` condition" (I think that needs to be "an `if` condition), but I expect that we would hit the `else` branch even if no `if` condition is attached to the loop, but the iterator is just empty (like `&[]`).

Reply:
Oh I didn't realize that could have a `else` branch without the `if`. Interesting. I will also need to update the `_did_loop` condition since it changed since #896.

Reply:
I'm actually not sure if you can, but I would expect them to be orthogonal.

Reply:
I confirm it works with just the `else`. It's quite crazy how many things I didn't know about askama. ^^'